crates/matrix-sdk/src/test_utils/mocks.rs

```diff
@@ -801,10 +868,286 @@ impl<'a> MockEndpoint<'a, RoomSendEndpoint> {
+    ///     .mock_room_send()
+    ///     .for_type("m.room.message".into())
+    ///     .ok(event_id)
+    ///     .expect(1)
+    ///     .mount()
+    ///     .await;
+    ///
+    /// let responseNotMocked = room.send_raw("m.room.reaction", json!({ "body": "Hello world" })).await;
+    /// // The `m.room.reaction` event type should not be mocked by the server.
+    /// assert!(responseNotMocked.is_err());
+    ///
+    ///
+    /// let response = room.send_raw("m.room.message", json!({ "body": "Hello world" })).await?;
+    /// // The `m.room.message` event type should be mocked by the server.
+    /// assert_eq!(
+    ///     event_id,
+    ///     response.event_id,
+    ///     "The event ID we mocked should match the one we received when we sent the event"
+    /// );
+    /// # anyhow::Ok(()) });
+    /// ```
+    pub fn for_type(self, event_type: MessageLikeEventType) -> Self {
+        Self {
+            mock: self
+                .mock
+                .and(path_regex(format!(r"^/_matrix/client/v3/rooms/.*/send/{}", event_type))),
+            ..self
+        }
+    }
+
+    /// Returns a send endpoint that emulates success, i.e. the event has been
+    /// sent with the given event id.
+    ///
+    /// # Examples
+    /// ```
+    /// # tokio_test::block_on(async {
+    /// use matrix_sdk::{ruma::{room_id, event_id}, test_utils::mocks::MatrixMockServer};
+    /// use serde_json::json;
+    ///
+    /// let mock_server = MatrixMockServer::new().await;
+    /// let client = mock_server.client_builder().build().await;
+    ///
+    /// mock_server.mock_room_state_encryption().plain().mount().await;
+    ///
+    /// let room = mock_server
+    ///     .sync_joined_room(&client, room_id!("!room_id:localhost"))
+    ///     .await;
+    ///
+    /// let event_id = event_id!("$some_id");
+    /// let send_guard = mock_server
+    ///     .mock_room_send()
+    ///     .ok(event_id)
+    ///     .expect(1)
+    ///     .mount_as_scoped()
+    ///     .await;
+    ///
+    /// let response = room.send_raw("m.room.message", json!({ "body": "Hello world" })).await?;
+    ///
+    /// assert_eq!(
+    ///     event_id,
+    ///     response.event_id,
+    ///     "The event ID we mocked should match the one we received when we sent the event"
+    /// );
+    /// # anyhow::Ok(()) });
+    /// ```
+    pub fn ok(self, returned_event_id: impl Into<OwnedEventId>) -> MatrixMock<'a> {
+        self.ok_with_event_id(returned_event_id.into())
+    }
+}
+
+/// A prebuilt mock for sending a state event in a room.
+#[derive(Default)]
+pub struct RoomSendStateEndpoint {
+    state_key: Option<String>,
+    event_type: Option<StateEventType>,
+}
+
+impl<'a> MockEndpoint<'a, RoomSendStateEndpoint> {
+    fn generate_path_regex(endpoint: &RoomSendStateEndpoint) -> String {
+        format!(
+            r"^/_matrix/client/v3/rooms/.*/state/{}/{}",
+            endpoint.event_type.as_ref().map(|t| t.to_string()).unwrap_or(".*".to_owned()),
```

Comment:
Not that we care too much about optimizing this code, but

```suggestion
            endpoint.event_type.as_ref().map(|t| t.to_string()).unwrap_or_else(|| ".*".to_owned()),
```